### PR TITLE
Add ThreadLocalWithFallback

### PR DESCRIPTION
**Pull request by spkrka**
This prevents needlessly duplicating too many
internal objects inside ThreadLocalSparkeyReader
when it's being used from ForkJoinPool.commonPool
in Java17+

### CODE REVIEW

File: src/main/java/com/spotify/sparkey/extra/ThreadLocalWithFallback.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+ * when being called from ForkJoinPool.commonPool().
+ * The common pool will clear all values for ThreadLocal after finishing a task,
+ * so it can not be used for retaining expensive objects on the same thread across
+ * multiple tasks.
+ *
+ * This subclass will instead let values survive by using a ConcurrentHashMap
+ * based on Thread Id which is stable as long as the thread lives. This map is also cleaned up by
+ * detecting thread death using a WeakReference.
+ */
+class ThreadLocalWithFallback<T> extends ThreadLocal<T> {
```
Comment (caesar-ralf) [inline src/main/java/com/spotify/sparkey/extra/ThreadLocalWithFallback.java:37]: Nitty-pick: I think the name could be something more related to this `ThreadLocal` storing the same value while the thread is alive.

I was thinking something like `PersistentThreadLocal`. What do you think? Again, nitty-pick comment. Feel free to ignore

File: src/main/java/com/spotify/sparkey/extra/ThreadLocalWithFallback.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    final long threadId = thread.getId();
+    final T fallbackValue = fallback.get(threadId);
+    if (fallbackValue != null) {
+      // This case is typically only reachable on Java 17+ using ForkJoinPool.commonPool()
+      return fallbackValue;
+    }
+
+    final T value = Objects.requireNonNull(supplier.get());
+    setFallback(value, thread, threadId);
+    return value;
```
Comment (caesar-ralf) [inline src/main/java/com/spotify/sparkey/extra/ThreadLocalWithFallback.java:63]: nitty-pick: Can be simplified with `computeIfAbsent`

```suggestion
    return fallback.computeIfAbsent(tid -> {
      final T value = Objects.requireNonNull(supplier.get());
      setFallback(value, thread, threadId);
      return value;
    });
```

File: src/main/java/com/spotify/sparkey/extra/ThreadLocalWithFallback.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+  public static <S> ThreadLocalWithFallback<S> withInitial(Supplier<? extends S> supplier, Consumer<? super S> closer) {
+    return new ThreadLocalWithFallback<>(supplier, closer);
+  }
+
+  @Override
+  public void set(T value) {
+    final Thread thread = Thread.currentThread();
+    final long threadId = thread.getId();
+    setFallback(value, thread, threadId);
```
Comment (caesar-ralf) [inline src/main/java/com/spotify/sparkey/extra/ThreadLocalWithFallback.java:89]: should it have a `Objects.requireNonNull` check here as well?

File: src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import static org.junit.Assert.assertTrue;
+
+public class ThreadLocalWithFallbackTest {
+
+  private final AtomicInteger closedThreads = new AtomicInteger();
+  private final ThreadLocalWithFallback<Long> threadLocal =
+          ThreadLocalWithFallback.withInitial(() -> Thread.currentThread().getId(), threadId -> closedThreads.incrementAndGet());
+
+  @Test
+  public void testSimple() {
```
Comment (caesar-ralf) [inline src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java:18]:
```suggestion
  public void testReturnsMainThreadIdWhenInMainThread() {
```

File: src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+import org.junit.Test;
+
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class ThreadLocalWithFallbackTest {
```
Comment (caesar-ralf) [inline src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java:11]: missing a test when you try to set initial value to a `null` value

File: src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  public void testForkJoinPoolCommonPool() {
+    final int poolSize = ForkJoinPool.commonPool().getParallelism();
+    final AtomicInteger failures = new AtomicInteger();
+    for (int i = 0; i < 10000; i++) {
+      int finalI = i;
+      ForkJoinPool.commonPool().execute(() -> {
+        checkFailure(failures);
+        if (finalI < poolSize) {
+          try {
+            Thread.sleep(1);
```
Comment (caesar-ralf) [inline src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java:41]: out of curiosity, why are you doing a `Thread.sleep` here? Trying to force it to use the whole pool?

File: src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+          } catch (InterruptedException e) {
+          }
+        }
+      });
+    }
+    assertEquals(poolSize, threadLocal.fallbackSize());
```
Comment (caesar-ralf) [inline src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java:47]: couldn't it be potentially `poolSize + 1` because commonpool can end up using the main thread for optimization? Wondering as this test might fail sometimes because of that

**Review by caesar-ralf — APPROVED** [review]
LGTM, I have nitty-pick comments that you can feel free to ignore

File: src/main/java/com/spotify/sparkey/extra/ThreadLocalWithFallback.java (hunk 1 above)
Comment (spkrka) [reply src/main/java/com/spotify/sparkey/extra/ThreadLocalWithFallback.java:37]: Yes, will think about this.

File: src/main/java/com/spotify/sparkey/extra/ThreadLocalWithFallback.java (hunk 2 above)
Comment (spkrka) [reply src/main/java/com/spotify/sparkey/extra/ThreadLocalWithFallback.java:63]: Might be nice to avoid the extra object churn since the lambda references values on the outside (supplier, thread, threadId)

File: src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java (hunk 5 above)
Comment (spkrka) [reply src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java:11]: Right, will think about what is the correct behavior in that case

File: src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java (hunk 6 above)
Comment (spkrka) [reply src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java:41]: Yes, exactly

File: src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java (hunk 7 above)
Comment (spkrka) [reply src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java:47]: I guess the test should be rewritten if that happens, but not needed right now

Comment (spkrka) [reply src/test/java/com/spotify/sparkey/extra/ThreadLocalWithFallbackTest.java]: I added a check for main thread
